### PR TITLE
Specflow mars

### PR DESCRIPTION
**Pull request by ayshafirdo**
added files for specflow for language and skills

### CODE REVIEW

File: SpecFlowProjectMars1/Features/LanguageSkills.feature [hunk 1]
```diff
@@ -0,0 +1,67 @@
+﻿Feature: Manage Languages
+  As a user
+  I want to add new languages to my profile
+  So that I can showcase the languages I know
+ @language
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Features/LanguageSkills.feature]: please separate feature files into two files skills and language

File: SpecFlowProjectMars1/Features/LanguageSkills.feature [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  Scenario Outline: Add a new language
+    Given the user logs into ProjectMars
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Features/LanguageSkills.feature]: Please increase your coverage around the features you  are testing here is an idea of what is missing 

• Negative testing with valid input
• Negative testing with invalid input
• Destructive testing

E.g of above scenarios
Happy path tests check basic functionality and the acceptance criteria of the API. We later extend positive tests to include optional parameters and extra functionality.

The next group of tests is negative testing where we expect the application to gracefully handle problem scenarios with both valid user input (for example, trying to add an existing username) and invalid user input (trying to add a username which is null).

Destructive testing is a deeper form of negative testing where we intentionally attempt to break the API to check its robustness (for example, sending a huge payload body in an attempt to overflow the system).

File: SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    [Binding]
+    public class ManageLanguagesStepDefinitions:CommonDriver
+    {
+        [Given(@"the language ""([^""]*)"" with level ""([^""]*)"" is already present")]
+        public void GivenTheLanguageWithLevelIsAlreadyPresent(string language, string level)
+        {
+            WebDriverWait wait = new WebDriverWait(driver, TimeSpan.FromSeconds(10));
+
+            // Navigate to the languages section if not already there
+            IWebElement languageSectionButton = driver.FindElement(By.XPath("//a[contains(text(),'Languages')]"));
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs]: you steps should not have any webelements, they should be kept in pages

File: SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            languageSectionButton.Click();
+
+            // Wait for the languages section to load
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs]: you are missing asserts please add them

File: SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            wait.Until(d => d.FindElement(By.XPath("//h3[text()='Languages']")));
+
+            try
+            {
+                // Check if the language with the specified level is already in the list
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs]: 2 State and Clean up 


example before adding language you need to make sure there are no existing languages as this could result in false positive test pass
you will need to think about this for all the scenarios, this should be done before you start any tests 

the state is say that you have a test that is adding a language english, the second time you run this test you should delete any existing language so that the state is reset, this will guarantee that your test does not produce false negatives and passes always. You can not assume that test execution order is not always the same, imagine if you run the tests  in parallel mode the all tests will run at the same time so you need to also make sure you do not wipe out other tests data
hence why when you remove data that other tests are using.  The best thing to do is delete only data you added for the test to be correct
e.g test 1 added english and french as language test 2 added hindi then when you reset data you already know what language you are adding for each test
there for on reset of the two test you should delete english and french for test 1 and hindi for test 2

So to sum up you should delete all data before the start of any test and then after each test you should delete the data that a particular test is adding!
This is ensure that upon staring the test pack you have reset state and upon finishing each test deletes their data that they added!
Think about first clean up before any test runs (you delete all data around this feature e.g  before any language test run you delete all languages). Then after each test delete only data that the test has added! (test1 added english and test2 to added french then test1 should only delete english and test2 should delete french only) 

also say you are updating a language then you must added language then update it do not assume the language is there! This applies for delete or any scenario

File: SpecFlowProjectMars1/Pages/Skill.cs [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SpecFlowProjectMars1.Pages
+{
+    public class Skill
+    {
+        public void AddNewSkill(IWebDriver driver, string skillName, string skillLevel)
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Pages/Skill.cs]: 5/
Please implement the POM Pattern, please see the below 
https://docs.specflow.org/projects/specflow/en/latest/ui-automation/Selenium-with-Page-Object-Pattern.html

File: SpecFlowProjectMars1/Utilities/CommonDriver.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SpecFlowProjectMars1.Utilities
+{
+    public class CommonDriver
+    {
+       public IWebDriver driver;
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Utilities/CommonDriver.cs]: finish implementing this 

4/
Please do not pass the driver In action methods (methods which in pages doing things like create, delete etc ) instead get the pages to extend to common driver where you will create a common driver class which will initialise the web driver 

public class CommonDriver
    {
        public static IWebDriver driver;

        public void initialise()
        {
            // create driver
            driver = new ChromeDriver();

        }

    }

File: SpecFlowProjectMars1/Hooks/Hooks.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+            LoginActions loginPageobj = new LoginActions();
+            loginPageobj.Login(driver, "test123456@test.com", "test123456");
+            ClearSkillTestData(driver);
+        }
+
+        [AfterScenario("language")]
+        public void AfterScenarioLanguage()
+        {
+            ClearLanguageTestData(driver);
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Hooks/Hooks.cs]: this is still not done correctly please see comment in step defs

File: SpecFlowProjectMars1/StepDefinitions/LanguageSkills.cs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+using SeleniumExtras.WaitHelpers;
+
+namespace SpecFlowProjectMars1.StepDefinitions
+{
+    [Binding]
+    public class LanguageSkills:CommonDriver
+    {
+        private IWebDriver driver;
+        private WebDriverWait wait;
+        Language languageobj = new Language();
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/StepDefinitions/LanguageSkills.cs]: Please make use of constructors, please see 
 https://www.tutorialspoint.com/csharp/csharp_classes.htm

https://www.knowledgehut.com/tutorials/csharp/csharp-constructors

File: SpecFlowProjectMars1/StepDefinitions/LanguageSkills.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        Skill skillobj=new Skill();
+
+        [Given(@"the user logs into ProjectMars")]
+        public void GivenTheUserLogsIntoProjectMars()
+        {
+            driver = new ChromeDriver();
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/StepDefinitions/LanguageSkills.cs]: this is already initialised in common driver why is this here ?

File: SpecFlowProjectMars1/StepDefinitions/LanguageSkills.cs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        [When(@"I update ""([^""]*)"" to ""([^""]*)""")]
+        public void WhenIUpdateTo(string oldLang, string newLang)
+        {
+            languageobj.EditLanguage(driver,oldLang,newLang);
+        }
+
+        [Then(@"""([^""]*)"" is added to list")]
+        public void ThenIsAddedToList(string newLang)
+        {
+            languageobj.VerifyLanguageUpdated(driver,newLang);
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/StepDefinitions/LanguageSkills.cs]: Please do not pass the driver In action methods (methods which in pages doing things like create, delete etc ) instead get the pages to extend to common driver where you will create a common driver class which will initialise the web driver

**Review by amanbansalnz — CHANGES_REQUESTED** [review]
I have reviewed your task please see the following feedback below

File: SpecFlowProjectMars1/Features/LanguageSkills.feature (hunk 1 above)
Comment (ayshafirdo) [reply SpecFlowProjectMars1/Features/LanguageSkills.feature]: ok will separate and update

File: SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs (hunk 3 above)
Comment (ayshafirdo) [reply SpecFlowProjectMars1/StepDefinitions/ManageLanguagesStepDefinitions.cs]: missed to remove this file..it was not used anywhere..sorry will remove it

File: SpecFlowProjectMars1/Utilities/CommonDriver.cs (hunk 7 above)
Comment (ayshafirdo) [reply SpecFlowProjectMars1/Utilities/CommonDriver.cs]: okay will implement this

File: SpecFlowProjectMars1/Hooks/Hooks.cs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        {
+            LoginActions loginPageobj = new LoginActions();
+            loginPageobj.Login("test123456@test.com", "test123456");
+            ClearSkillTestData();
+        }
+
+        [AfterScenario("language")]
+        public void AfterScenarioLanguage()
+        {
+            ClearLanguageTestData();
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Hooks/Hooks.cs]: so this you need to clean only the test data the test is adding, e.g test1 added english then you need to delete english

File: SpecFlowProjectMars1/Hooks/Hooks.cs [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+            LoginActions loginPageobj = new LoginActions();
+            loginPageobj.Login("test123456@test.com", "test123456");
+            ClearLanguageTestData();
+        }
+        [BeforeScenario("skill")]
+        public void BeforeScenarioSkill()
+        {
+            LoginActions loginPageobj = new LoginActions();
+            loginPageobj.Login("test123456@test.com", "test123456");
+            ClearSkillTestData();
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Hooks/Hooks.cs]: this should not be in this hook but in another hook [BeforeTestRun] [AfterTestRun]
 As the name suggests “BeforeTestRun” will trigger before the test run and “AfterTestRun” will trigger after complete test run.look at https://codebun.com/how-to-implement-hooks-in-specflow-c/

File: SpecFlowProjectMars1/Utilities/ElementHelper.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+using OpenQA.Selenium;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SpecFlowProjectMars1.Utilities
+{
+    public class ElementHelper : CommonDriver
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Utilities/ElementHelper.cs]: remove code if not in use

**Review by amanbansalnz — CHANGES_REQUESTED** [review]
please fix cleanup

File: SpecFlowProjectMars1/Hooks/Hooks.cs [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+using SpecFlowProjectMars1.Utilities;
+using TechTalk.SpecFlow;
+
+namespace SpecFlowProjectMars1.Hooks1
+{
+
+
+    [Binding]
+    public class Hooks : CommonDriver
+    {
```
Comment (amanbansalnz) [inline SpecFlowProjectMars1/Hooks/Hooks.cs:15]: hey why have you added everything in the hooks class?

File: SpecFlowProjectMars1/Hooks/Hooks.cs (hunk 8 above)
Comment (amanbansalnz) [reply SpecFlowProjectMars1/Hooks/Hooks.cs]: https://docs.specflow.org/projects/specflow/en/latest/Bindings/Hooks.html


https://codebun.com/how-to-implement-hooks-in-specflow-c/

**Review by amanbansalnz — APPROVED** [review]
I am happy to approve
